cilium-dbg/cmd/shell.go

```diff
@@ -0,0 +1,179 @@
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/cilium/hive/script"
+	"github.com/spf13/cobra"
+	"golang.org/x/term"
+
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/version"
+)
+
+var shellCmd = &cobra.Command{
+	Use:   "shell [command] [args]...",
+	Short: "Connect to the Cilium shell",
+	Run:   shell,
+}
+
+var stdReadWriter = struct {
+	io.Reader
+	io.Writer
+}{
+	Reader: os.Stdin,
+	Writer: os.Stdout,
+}
+
+func init() {
+	RootCmd.AddCommand(shellCmd)
+}
+
+func dialShell() (net.Conn, error) {
+	var conn net.Conn
+	until := time.Now().Add(time.Minute)
+	for {
+		var err error
+		conn, err = net.Dial("unix", defaults.ShellSockPath)
+		if err == nil {
+			break
+		}
+		if time.Now().After(until) {
+			return nil, fmt.Errorf("Timed out dialing %q\n", defaults.ShellSockPath)
+		}
+		fmt.Fprintf(os.Stderr, "Dial(%q): %s. Retrying...\n", defaults.ShellSockPath, err)
+		time.Sleep(time.Second)
+	}
+	return conn, nil
+}
+
+func shellExchange(w io.Writer, format string, args ...any) error {
+	conn, err := dialShell()
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	_, err = fmt.Fprintf(conn, format+"\nexit\n", args...)
+	if err != nil {
+		return err
+	}
+	bio := bufio.NewReader(conn)
+	for {
+		line, _, err := bio.ReadLine()
+		if err != nil {
+			return nil
+		}
+		switch string(line) {
+		case "<<end>>":
+			return nil
+		case "[stdout]":
+		default:
+			w.Write(line)
+			w.Write(newline)
+		}
+	}
+
+}
+
+func shell(cmd *cobra.Command, args []string) {
+	if len(args) > 0 {
+		shellExchange(os.Stdout, "%s", strings.Join(args, " "))
+	} else {
+		interactiveShell()
+	}
+}
+
+func interactiveShell() {
+	// Try to set the terminal to raw mode (so that cursor keys work etc.)
+	restore, err := script.MakeRaw(0)
+	if err != nil {
+		panic(err)
+	}
+	defer restore()
+
+	// Listen for SIGINT to break.
+	sigs := make(chan os.Signal, 1)
+	defer signal.Stop(sigs)
+	signal.Notify(sigs, os.Interrupt)
+
+	term := term.NewTerminal(stdReadWriter, "cilium> ")
+	printShellGreeting(term)
+
+	for {
+		// Try to dial the shell.sock. Since it takes a moment for the agent to come up and this
+		// is meant for interactive use we'll try to be helpful and retry the dialing until
+		// agent comes up.
+		conn, err := dialShell()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+			os.Exit(1)
+		}
+
+		stop := make(chan struct{})
+		go func() {
+			select {
+			case <-stop:
+			case <-sigs:
+				// On interrupt close the connection.
+				conn.Close()
+			}
+		}()
+
+		bio := bufio.NewReader(conn)
+
+	outer:
+		for {
+			line, err := term.ReadLine()
+			if err != nil {
+				return
+			}
+			if _, err = fmt.Fprintln(conn, line); err != nil {
+				break
+			}
+
+			// Read until the command finishes with <<end>>.
+			for {
+				line, _, err := bio.ReadLine()
+				if err != nil {
+					break outer
```

Comment:
can we return and defer close(stop) here instead of using the break block?

Reply:
huh weird, why didn't I write it that way in the first place...

Reply:
doh, because we don't want to return, but we want to redial the connection instead.